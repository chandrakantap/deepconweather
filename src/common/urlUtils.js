export function getQueryParams(searchString = "") {
  return searchString
    .slice(1)
    .split("&")
    .map((p) => p.split("="))
    .reduce((obj, [key, value]) => ({ ...obj, [key]: value }), {});
}
