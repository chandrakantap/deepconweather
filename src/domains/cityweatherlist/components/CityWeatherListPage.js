import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { createSelector } from "reselect";
import { Link } from "react-router-dom";
import Throbber from "common/ui/Throbber";
import SingleCityInfo from "domains/cityweatherlist/components/singlecityInfo/SingleCityInfo";
import CitySearch from "domains/citysearch/components/CitySearch";
import {
  loadCityListAction,
  toggleFavouriteAction,
  removeCityAction,
  addCityToListAction,
} from "domains/cityweatherlist/store/cityWeatherListPageActions";
import {
  cityListSelector,
  isCityListLoadedSelector,
} from "domains/cityweatherlist/store/cityWeatherListPageSelectors";
import styles from "./CityWeatherListPage.module.css";

const selector = createSelector(
  cityListSelector,
  isCityListLoadedSelector,
  (cityList, isCityListLoaded) => ({
    cityList,
    isCityListLoaded,
  })
);

function CityWeatherListPage() {
  const dispatch = useDispatch();
  const { cityList, isCityListLoaded } = useSelector(selector);

  useEffect(() => {
    dispatch(loadCityListAction());
  }, [dispatch]);

  const onClickFavourite = (city) => {
    dispatch(toggleFavouriteAction(city.id));
  };
  const onClickRemove = (city) => {
    dispatch(removeCityAction(city.id));
  };
  const addNewCityToList = (city) => {
    dispatch(addCityToListAction(city));
  };

  if (!isCityListLoaded) {
    return (
      <section>
        <Throbber />
      </section>
    );
  }

  return (
    <section>
      <header className={styles.header}>
        <h1>City Weather Info</h1>
      </header>
      <main className={styles.cityList}>
        <div className={styles.citySearchContainer}>
          <CitySearch onSelect={addNewCityToList} />
        </div>
        {cityList.map((city) => (
          <Link
            to={`/detail?cityId=${city.id}&cityName=${city.name}&region=${city.region}&country=${city.country}`}
            key={city.id}
          >
            <SingleCityInfo
              city={city}
              onClickFavourite={onClickFavourite}
              onClickRemove={onClickRemove}
            />
          </Link>
        ))}
      </main>
    </section>
  );
}

export default CityWeatherListPage;
