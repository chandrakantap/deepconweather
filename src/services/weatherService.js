const apiBaseURL = 'http://api.weatherstack.com';
const accessKey = process.env.REACT_APP_WS_API_KEY;

async function getCityWeather(cityName) {
    const response = await fetch(`${apiBaseURL}/current?access_key=${accessKey}&query=${cityName}`);
    const { location: { name, country }, current } = await response.json();
    return {
        name,
        country,
        current
    }
}

export default { getCityWeather }