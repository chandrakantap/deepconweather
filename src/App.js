import React from 'react';
import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import {
  BrowserRouter as Router,
  Switch,
  Route
} from 'react-router-dom';
import { Provider } from 'react-redux';
import rootReducer from 'store/rootReducer';

import CityWeatherListPage from 'domains/cityweatherlist/components/CityWeatherListPage';
import CityWeatherDetailPage from 'domains/weatherdetail/components/CityWeatherDetailPage';
import GeoLocationAwareRedirector from 'domains/geolocation/components/GeoLocationAwareRedirector';

const middlewares = [thunk];
const composeEnhancers = window['__REDUX_DEVTOOLS_EXTENSION_COMPOSE__'] || compose;
const store = createStore(rootReducer, composeEnhancers(applyMiddleware(...middlewares)));

function App({ children }) {
  return (
    <Provider store={store}>
      <Router>
        <Switch>
          <Route exact path="/" component={GeoLocationAwareRedirector} />
          <Route exact path="/list" component={CityWeatherListPage} />
          <Route exact path="/detail/:cityname/:country" component={CityWeatherDetailPage} />
        </Switch>
      </Router>
    </Provider>
  );
}

export default App;
