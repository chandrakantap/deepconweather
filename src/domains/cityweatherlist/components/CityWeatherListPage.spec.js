import React from 'react';
import {
    fireEvent,
    waitForDomChange,
    screen, cleanup
} from '@testing-library/react';
import { render } from 'common/test-utils';
import CityWeatherListPage from './CityWeatherListPage';
import App from 'App';
import weatherService from 'services/weatherService';

describe('CityWeatherListPage', () => {
    let getCityWeatherSpy;
    beforeAll(() => {
        getCityWeatherSpy = jest.spyOn(weatherService, 'getCityWeather').mockImplementation((cityName) => {
            const cityData = cityName.split(',');
            return Promise.resolve({
                "name": cityData[0],
                "country": cityData[1],
                "current": {
                    "observation_time": "02:21 PM",
                    "temperature": 32,
                    "weather_code": 113,
                    "weather_icons": [
                        "https://assets.weatherstack.com/images/wsymbols01_png_64/wsymbol_0001_sunny.png"
                    ],
                    "weather_descriptions": [
                        "Sunny"
                    ],
                    "wind_speed": 17,
                    "wind_degree": 250,
                    "wind_dir": "WSW",
                    "pressure": 1006,
                    "precip": 0,
                    "humidity": 75,
                    "cloudcover": 0,
                    "feelslike": 31,
                    "uv_index": 8,
                    "visibility": 6,
                    "is_day": "yes"
                }
            });
        })
    });
    afterEach(() => {
        cleanup();
    });
    afterAll(() => {
        getCityWeatherSpy.restore();
    })

    test('should display 15 cities by default', async () => {
        const { container } = render(<CityWeatherListPage />);
        await waitForDomChange();
        expect(container.querySelectorAll('main.cityList section.root').length).toBe(15);
    });

    test('should display cities from localStorage if present default', async () => {
        window.localStorage.setItem(
            'LIST_PAGE_CITIES_SK',
            '[\
                {"name":"Tokyo,Japan","isFavourite":false},\
                {"name":"Delhi,India","isFavourite":true}\
            ]'
        );
        const { container } = render(<CityWeatherListPage />);
        await waitForDomChange();
        expect(container.querySelectorAll('main.cityList section.root').length).toBe(2);
    });

    test('should delete cities on click delete icon', async () => {
        window.localStorage.setItem(
            'LIST_PAGE_CITIES_SK',
            '[\
                {"name":"Tokyo,Japan","isFavourite":false},\
                {"name":"Delhi,India","isFavourite":true},\
                {"name":"Shanghai,China","isFavourite":false},\
                {"name": "Mexico City,Mexico", "isFavourite": false}\
            ]'
        );
        const { container } = render(<CityWeatherListPage />);
        await waitForDomChange();
        fireEvent.click(screen.queryByTestId('Tokyo_Japan_deleteBtn'));
        expect(container.querySelectorAll('main.cityList section.root').length).toBe(3);
    });

    test('should set favourite on click favourite icon', async () => {
        window.localStorage.setItem(
            'LIST_PAGE_CITIES_SK',
            '[\
                {"name":"Tokyo,Japan","isFavourite":false}\
            ]'
        );
        const { container } = render(<CityWeatherListPage />);
        await waitForDomChange();
        fireEvent.click(screen.queryByTestId('Tokyo_Japan_favBtn'));
        expect(container).toMatchSnapshot();
    });
})