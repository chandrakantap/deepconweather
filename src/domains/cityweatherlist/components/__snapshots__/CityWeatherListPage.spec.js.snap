// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CityWeatherListPage should set favourite on click favourite icon 1`] = `
<div>
  <section>
    <header
      class="header"
    >
      <h1>
        City Weather Info
      </h1>
      <button
        class="iconButton"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M.69 6.997A17.925 17.925 0 0 1 12 3c4.285 0 8.22 1.497 11.31 3.997l-1.256 1.556A15.933 15.933 0 0 0 12 5C8.191 5 4.694 6.33 1.946 8.553L.69 6.997zm3.141 3.89A12.946 12.946 0 0 1 12 8c3.094 0 5.936 1.081 8.169 2.886l-1.257 1.556A10.954 10.954 0 0 0 12 10c-2.618 0-5.023.915-6.912 2.442l-1.257-1.556zm3.142 3.89A7.967 7.967 0 0 1 12 13c1.904 0 3.653.665 5.027 1.776l-1.257 1.556A5.975 5.975 0 0 0 12 15c-1.428 0-2.74.499-3.77 1.332l-1.257-1.556zm3.142 3.89A2.987 2.987 0 0 1 12 18c.714 0 1.37.25 1.885.666L12 21l-1.885-2.334z"
            />
          </g>
        </svg>
      </button>
    </header>
    <main
      class="cityList"
    >
      <div
        class="citySearchContainer"
      >
        <div
          class="root"
        >
          <div
            class="root"
          >
            <input
              class="searchTxtFld"
              placeholder="Search for cities..."
              type="text"
              value=""
            />
            
          </div>
        </div>
      </div>
      <section
        class="root"
      >
        <a
          href="/detail?cityId=TOKYO_TOKYO_JAPAN&cityName=Tokyo&region=Tokyo&country=Japan"
        >
          <div>
            <h1
              class="cityName"
            >
              Tokyo
            </h1>
            <h2
              class="countryName"
            >
              Tokyo, Japan
            </h2>
          </div>
        </a>
        <h2
          class="temperature"
        >
          32
           â„ƒ
        </h2>
        <div>
          <button
            class="iconButton favouriteIcon"
            data-testid="TOKYO_TOKYO_JAPAN_favBtn"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55l-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
              />
            </svg>
          </button>
        </div>
        <button
          class="iconButton deleteIcon visibleOnHover"
          data-testid="TOKYO_TOKYO_JAPAN_deleteBtn"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </button>
      </section>
    </main>
  </section>
</div>
`;
